module/zfs/sa.c

```diff
@@ -369,7 +370,7 @@ sa_attr_op(sa_handle_t *hdl, sa_bulk_attr_t *bulk, int count,
 			if (bulk[i].sa_data) {
 				SA_COPY_DATA(bulk[i].sa_data_func,
 				    bulk[i].sa_addr, bulk[i].sa_data,
-				    bulk[i].sa_size);
+				    MIN(bulk[i].sa_size, bulk[i].sa_length));
```

Comment:
Rather than successfully returning a truncated buffer perhaps we should return something like `EOVERFLOW` in this case.

Reply:
I'm not sure how much it'd help though without re-doing more of the interface to let the caller know just how much space is needed. When the buffer is too big, it also doesn't tell you how much of the buffer is actually valid, so there's already at least an implicit notion of 'the caller should know what to do with the data its given'.

`sa_lookup()` should probably be changed, but that's a much more extensive set of work given all of the callers.

Reply:
That's a fair point.  Okay let's fix the overflow now and we can always revisit this interface as needed.